src/backtracking/hamiltonian_cycle.rs

```diff
@@ -0,0 +1,248 @@
+//! This module provides functionality to find a Hamiltonian cycle in a graph.
+//! Source: [Wikipedia](https://en.wikipedia.org/wiki/Hamiltonian_path_problem)
+
+/// Represents potential errors when working with an adjacency matrix.
+#[derive(Debug, PartialEq, Eq)]
+pub enum AdjMatError {
+    /// Indicates that the adjacency matrix is empty.
+    EmptyMat,
+    /// Indicates that the adjacency matrix is not square.
+    ImproperMat,
+    /// Indicates that the starting vertex is out of bounds.
+    StartOutOfBound,
+}
+
+/// Represents a graph using an adjacency matrix.
+struct Graph {
+    /// The adjacency matrix representing the graph.
+    adjacency_matrix: Vec<Vec<bool>>,
+}
+
+impl Graph {
+    /// Creates a new graph with the provided adjacency matrix.
+    ///
+    /// # Arguments
+    ///
+    /// * `adjacency_matrix` - A square matrix where each element indicates
+    ///                        the presence (`true`) or absence (`false`) of an edge
+    ///                        between two vertices.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the graph if successful, or an `AdjMatError` if there is an issue with the matrix.
+    fn new(adjacency_matrix: Vec<Vec<bool>>) -> Result<Self, AdjMatError> {
+        // Check if the adjacency matrix is empty.
+        if adjacency_matrix.is_empty() {
+            return Err(AdjMatError::EmptyMat);
+        }
+
+        // Validate that the adjacency matrix is square.
+        if adjacency_matrix
+            .iter()
+            .any(|row| row.len() != adjacency_matrix.len())
+        {
+            return Err(AdjMatError::ImproperMat);
+        }
+
+        Ok(Self { adjacency_matrix })
+    }
+
+    /// Returns the number of vertices in the graph.
+    fn num_vertices(&self) -> usize {
+        self.adjacency_matrix.len()
+    }
+
+    /// Determines if it is safe to include vertex `v` in the Hamiltonian cycle path.
+    ///
+    /// # Arguments
+    ///
+    /// * `v` - The index of the vertex being considered.
+    /// * `path` - A reference to the current path being explored.
+    /// * `pos` - The position of the current vertex being considered.
+    ///
+    /// # Returns
+    ///
+    /// `true` if it is safe to include `v` in the path, `false` otherwise.
+    fn is_safe(&self, v: usize, path: &[usize], pos: usize) -> bool {
+        // Check if the current vertex and the last vertex in the path are adjacent.
+        if !self.adjacency_matrix[path[pos - 1]][v] {
+            return false;
+        }
+
+        // Check if the vertex has already been included in the path.
+        !path[..pos].contains(&v)
```

Comment:
This is quite inefficient - $O(n)$. Equivalent computation could be done in constant time, when there is some `visited` variable being a hash set.

Reply:
I will refactor the logic to make it more efficient.